specs/eventing/control-plane.md

```diff
@@ -0,0 +1,985 @@
+send some additional events to the Trigger's `spec.subscriber` after the
+deletion.
+
+### Channel Lifecycle
+
+A Channel represents an Addressable endpoint (i.e. it MUST have a
+`status.address.url` field) which can receive, store, and forward events to
+multiple recipients (Subscriptions). Subscriptions are associated with a Channel
+based on the `spec.channel` field on the Subscription; it is expected that the
+controller for a Channel will also control the associated Subscriptions. When
+the Channel's `Ready` condition is `true`, the Channel MUST provide a
+`status.address.url` which accepts all valid CloudEvents and MUST forward the
+received events to each associated Subscription whose `Ready` condition is
+`true`. As described in the [Subscription Lifecycle](#subscription-lifecycle)
+section, a Channel MAY forward events to an associated Subscription which does
+not currently have a `true` `Ready` condition, including events received by the
+Channel before the `Subscription` was created.
+
+When a Channel is created, its `spec.channelTemplate` field MUST be populated to
+indicate which of several possible Channel implementations to use. It is
+RECOMMENDED to default the `spec.channelTemplate` field on creation if it is
+unpopulated. Once created, the `spec.channelTemplate` field MUST be immutable;
+the Channel MUST be deleted and re-created to change the `spec.channelTemplate`.
+This pattern is chosen to make it clear that changing `spec.channelTemplate` is
+not an atomic operation and that any implementation would be likely to result in
+event loss during the transition.
+
+### Subscription Lifecycle
+
+The lifecycle of a Subscription is independent of the channel it refers to in
+its `spec.channel` field. The `spec.channel` object reference may refer to
+either an `messaging.knative.dev/v1` Channel resource, or another resource which
+meets the `spec.subscribers` and `spec.delivery` required elements in the
+Channelable duck type. If the referenced `spec.channel` does not currently exist
+or its `Ready` condition is not `true`, then the Subscription's `Ready`
+condition MUST NOT be `true`, and the reason SHOULD indicate that the
+corresponding channel is missing or not ready.
+
+The Subscription MUST also set the `status.physicalSubscription` URIs by
+resolving the `spec.subscriber`, `spec.reply`, and
+`spec.delivery.deadLetterSink` as described in
+[Addressable resolution](#addressable-resolution) before setting the `Ready`
+condition to `true`. If any of the addressable fields fails resolution, the
+Subscription MUST set the `Ready` condition to `false`, and at least one
+condition should indicate the reason for the error. (It is acceptable for none
+of the `spec.subscriber`, `spec.reply`, and `spec.delivery.deadLetterSink`
+fields to contain a `ref` field.)
+
+At least one of `spec.subscriber` and `spec.reply` MUST be set; if only
+`spec.reply` is set, the behavior is equivalent to setting `spec.subscriber`
```

Comment:
Couple of things:
- I think we need a MUST in here, something like ".. the behavior MUST be equivalent.."
- I'm not following the purpose of this and I think we need to explain the reason for this magic.
- w.r.t. the Channel SHOULD NOT... , do you mean: if the Channel does not advertise the ability to process replies then the Subscription SHOULD NOT .... something???  Can't say for sure, but it feels like the SHOULD NOT here is backwards... on the the wrong resource.

Reply:
Just seems to me that a platform that appears to go out of it's way to be explicit w.r.t. to fields (e.g. will update the user's spec fields that are empty with default values just so the user knows what's going on) would also want to ensure that spec.subscriber has a value instead of having the user "just know" the "spec.reply" will be used.

Besides that, it seems to me that if any field on a Subscription resource would be required it would be the "spec.subscriber" one since that's the main purpose of this resource.

> if only spec.reply is set, the behavior is equivalent to setting spec.subscriber...

Set to what? if we mean "spec.reply's value, then let's say that. If we keep this magic at all.

> except that the Channel SHOULD NOT advertise the ability to process replies during the delivery.

At this point in the doc I don't think we've talked about "advertise" anything so I'm not sure what this means. If this is a forward reference then let's add an href to the docs for it. But even so, I don't grok it. Why can't it process replies if the user set a spec.reply address??

Reply:
hmm that raises an interesting situation.  Specs are not at 1.0 yet (officially), api version is v1, and then there's the ref impl code level (not sure what version # they're at). Seems to me that the "v1" on the api might have been premature unless we're telling the KTC that they're not allowed to disagree with certain changes if it means breaking the api's v1 status. I guess I should ask... does the "v1" on the API align with the "v1" of the spec? Or can the spec be at v1 but the api be at v2? That would seem odd since the spec define the api.

Reply:
From @n3wscott :

> I am :-1:  on this change. There could be reasons why you want a “write” only subscriber and rather that add modes or some hard thing to the subscription, you can force a push based subscription into write only mode (no response possible) by using the reply  field. We have said for a long time that reply is the wrong term but it is too late to change it

I'm not sure about "too late to change it", but I do think there's a valid point on wanting to prevent replies. On the other hand, a user could also simply not set `reply`, and the replies from the `subscriber` would be discarded because there is no destination for them.

Reply:
I'm not following this - can we look at removing this magic or is it too late?